### ISSUE TITLE
RFE: Capability monitoring 

### ISSUE THREAD

**Issue by jaymzh** [systemd/systemd#10236]
There was a discussion at ASG2018 on whether systemd could monitor what capabilities a service used.

This would be an option to systemd that would add a small BPF program that monitors what capabilities a service is using. This would help users to strip the right CAPs from their services.

There was discussion about doing similar for syscalls, exec dependencies etc, where BPF may be not sufficient or too slow (there aren't tracepoints for every syscall, for example). It was discussed that we could use the audit frame work for this by providing a feature for systemd to add the right audit rules to log all syscalls and/or execs for a specific progress.

In both cases the feature would default to off and be enabled by something like: `AuditCapablities=true` or `AuditSyscalls=true`. Ideally, systemd would store these (in a property?) in a format suitable for then passing to `CapabilityBoundingSet` and `SystemCallFilter`.

**Comment by topimiettinen** [comment]
Long time a go I started working on similar feature using systemtap. [This systemtap script](https://github.com/topimiettinen/systemd-settings-generator) can monitor a test run and generate most systemd resource control directives. At that time BPF couldn't be used for that but has this maybe changed?

Before taking the systemtap approach, I also proposed changes to kernel to support monitoring capabilities and other resource limit uses:
https://lkml.org/lkml/2016/7/11/276
https://lkml.org/lkml/2016/6/13/857
Something like that could be reconsidered too.

**Comment by poettering** [comment]
hmm, actually, i don't know how we'd implement this. We can't really know which caps the kernel was looking for when it fails a syscall, and even if we knew we couldn't really store that away as seccomp currently only does plain bpf, not ebpf...

selinux uses the audit subsystem for tracking down such issues, but for the caps case I am not sure how we'd make this work...

**Comment by williampsmith** [comment]
In the case of `capmond`, rather than using seccomp, we have our own eBPF program that attaches to the `cap_capable` [kernel function](https://elixir.bootlin.com/linux/latest/source/security/commoncap.c#L71) through kprobes.

**Comment by poettering** [comment]
@williampsmith ah, indeed that should work.

Hmm, but that's an internal kernel function, and not official userspace-facing, stable API. I think hooking into that with kprobes is fine for a specialized tool, but for something so generic as systemd I am not sure it's the right choice. i.e. i am pretty sure systemd should not become a debugger/tracer of the kernel, because that#s effectively what we'd become if we started to use kprobes in our regular codepaths...

@jaymzh hmm, do you know what precisely what the benefit would be of having this in systemd rather than suggesting people to use capmond? @williampsmith does capmond have a nice way to hook it into an ExecStart= line of a systemd unit file? i.e. can it operate on a specific process (and its children), and do so by forking itself in the background (á la `strace -D`)?

**Comment by poettering** [comment]
@williampsmith btw, what's the url of capmond? is there already a public repo for that? google doesn't find it...

**Comment by davidstrauss** [comment]
I also raised the idea of something like `AuditPaths=true` to help profile a service for filesystem namespace pruning or writing MAC rules (e.g. selinux or AppArmor). Systems like selinux have their own auditing tools, but they're very targeted at their own rule structures.

**Comment by williampsmith** [comment]
@poettering 

> does capmond have a nice way to hook it into an ExecStart= line of a systemd unit file?

No, at the moment capmond will trace all cap events on the host in which it runs. It collects data on the cgroup belonging to the cap event, and then separately resolves it to the associated service/container/etc, basically by reading extended attributes on the cgroup. For security reasons, we didn't want a model where the service owner could selectively turn off capmond for his/her service. 

That said, plugging capmond into a cgroup at service start time *is* something of interest. One way we're considering of doing this securely is by using cgroup specific bpf, which allows us to attach only to cap events that occur for processes within a specific cgroup. 

> what's the url of capmond? is there already a public repo for that?

We haven't yet put up a public repo of the project yet, but are working on that diligently and hope to make it publicly available fairly soon.

**Comment by topimiettinen** [comment]
@davidstrauss systemtap can be used to determine paths used and then that information can then be used to determine if for example `ProtectHome` can be enabled or to suggest some `InaccessiblePaths`. My little script can also check many other things, like if `MemoryDenyWriteExecute` is OK.

But I agree with @poettering that kernel probe interface is not very stable. Perhaps there should be a stackable LSM which uses the security module interface (relatively stable) to collect capability usage (and other) information and then pass that on via netlink socket. @williampsmith have you considered that approach?
